File: 2002/ALL VB .NET SAMPLES/VB.NET - Framework - Key VB.NET Benefits/FullTimeEmployee.vb
'Copyright (C) 2002 Microsoft Corporation
'All rights reserved.
'THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER
'EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED WARRANTIES OF
'MERCHANTIBILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.

'Requires the Trial or Release version of Visual Studio .NET Professional (or greater).

Option Strict On

' The FullTimeEmployee class inherits from the Employee class and
' therefore has all the properties, methods and events that Employee
' has. FullTimeEmployee also extends the Employee class by adding the
' ComputeAnnualLeave method and overriding the ComputeBonus method.
Public Class FullTimeEmployee
    Inherits Employee

    ' This constructor simply calls the default constructor in the base
    ' class.
    Public Sub New()
        MyBase.New()
    End Sub

    ' This constructor calls the parameterized constructor in the base
    ' class, passing it the employee name.
    Public Sub New(ByVal Name As String)
        MyBase.New(Name)
    End Sub

    ' The AnnualLeave property is measured in days, and is unique to the
    ' FullTimeEmployee class.
    Public ReadOnly Property AnnualLeave() As Integer
        Get
            Return ComputeAnnualLeave()
        End Get
    End Property

    ' This Salary property procedure provides the implementation
    ' for the Salary property that was declared but not
    ' implemented in the base class.
    Public Overrides Property Salary() As Decimal
        Get
            Return c_Salary
        End Get
        Set(ByVal Value As Decimal)
            If Value < 30000.0 Or Value > 500000.0 Then
                Throw New ArgumentOutOfRangeException( _
                     "Salary", _
                     "Full-time employee salary must be between " & _
                     "$30,000 and $500,000")
            Else
                c_Salary = Value
            End If
        End Set
    End Property

    ' By implementing the ComputeAnnualLeave method, this class is
    ' extending the base class, Employee. The method does not appear in
    ' the base class, nor in the PartTimeEmployee class, which is also
    ' derived from Employee.
    '
    ' The method computes how long the employee has been with the
    ' company and gives him 4 days leave for each quarter of service.
    Public Function ComputeAnnualLeave() As Integer
        ' Code to compute annual leave.
        ' We're feeling generous in this demo, so we're giving everyone
        ' six weeks.
        Return 42
    End Function

    ' Implements the ComputeBonus method for full-time employees.
    Public Overrides Function ComputeBonus() As Decimal
        ' Full-time employees get an annual bonus of 1% of their
        ' salary.
        '
        ' Because the compiler interprets 0.01 as a Long, you convert
        ' it to Decimal with the CType function. That way, your math
        ' calculation will meet the rules of Option Strict.
        '
        ' Option Strict On prevents you from accidentally narrowing a
        ' data type (coercing a Long into a Short, for example). You
        ' can set Option Strict On at the top of each file, or at the
        ' project level.
        Return c_Salary * CType(0.01, Decimal)
    End Function
End Class

File: 2003/Code Samples/Framework Samples/Framework - Scoping, Overloading, Overriding/FullTimeEmployee.cs
//Copyright (C) 2002 Microsoft Corporation
//All rights reserved.
//THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER
//EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED WARRANTIES OF
//MERCHANTIBILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
//Requires the Trial or Release version of Visual Studio .NET Professional (or greater).
// The FullTimeEmployee class inherits from the Employee class and
// therefore has all the properties, methods and events that Employee
// has. FullTimeEmployee also extends the Employee class by adding the
// ComputeAnnualLeave method and overriding the ComputeBonus method.

using System;

public class FullTimeEmployee: Employee
{
    // This constructor simply calls the default constructor in the base
    // class.
    public FullTimeEmployee() 
	{
    }

    // This constructor calls the parameterized constructor in the base
    // class, passing it the employee name.
    public FullTimeEmployee(string Name): base(Name)
	{
    }

    // The AnnualLeave property is measured in days, and is unique to the
    // FullTimeEmployee class.
    public int AnnualLeave
	{
        get {
            return ComputeAnnualLeave();
        }
    }

    // This Salary property procedure provides the implementation
    // for the Salary property that was declared but not
    // implemented in the base class.
    public new decimal Salary
	{
        get 
		{
            return c_Salary;
        }
        set
		{
			if ((value < Convert.ToDecimal(30000.0)) || (value > Convert.ToDecimal(500000.0)))
			{
				throw new ArgumentOutOfRangeException("Salary",
					"Full-time employee salary must be between " +
					"$30,000 and $500,000");
			}
			else 
			{
				c_Salary = value;
			}
        }
    }

    // By implementing the ComputeAnnualLeave method, this class is
    // extending the base class, Employee. The method does not appear in
    // the base class, nor in the PartTimeEmployee class, which is also
    // derived from Employee.
    //
    // The method computes how long the employee has been with the
    // company and gives him 4 days leave for each quarter of service.
    public int ComputeAnnualLeave() 
	{
        // Code to compute annual leave.
        // We're feeling generous in this How-To, so we're giving everyone
        // six weeks.
        return 42;
    }

    // : the ComputeBonus method for full-time employees.
    public override decimal ComputeBonus() 
	{
        // Full-time employees get an annual bonus of 1% of their
        // salary.
        //
        // Because the compiler interprets 0.01 a Long, you convert
        // it to Decimal with the CType function. That way, your math
        // calculation will meet the rules of Option Strict.
        //
        //  prevents you from accidentally narrowing a
        // data type (coercing a long into a Short, for example). You
        // can set  at the top of each file, or at the
        // project level.
        return c_Salary * Convert.ToDecimal(0.01);
    }
}

